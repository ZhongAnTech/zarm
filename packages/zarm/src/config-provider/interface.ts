import type { MountContainer, ScrollContainer } from '../utils/dom';
import type { CssVars } from './cssVars';
import localeCN from './locale/zh_CN';

export type Locale = typeof localeCN;

export interface ConfigProviderProps {
  prefixCls?: string;
  locale?: Locale;
  primaryColor?: string;
  theme?: 'dark' | 'light';
  safeArea?: boolean;
  mountContainer?: MountContainer;
  scrollContainer?: ScrollContainer;
  cssVars?: CssVars;
}
