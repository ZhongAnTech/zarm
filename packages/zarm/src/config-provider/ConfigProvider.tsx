import React, { PureComponent, createContext } from 'react';
import type { Context } from 'react';
import type { ConfigProviderProps, Locale } from './PropsType';
import setTheme from './setTheme';
import setPrimaryColor from './setPrimaryColor';

const defaultConfig: ConfigProviderProps = {
  locale: ({} as unknown) as Locale,
  theme: 'light',
  primaryColor: '#00bc70',
};

export const LocaleContext: Context<Locale> = createContext(defaultConfig.locale);
export const ConfigContext: Context<ConfigProviderProps> = createContext(defaultConfig);

let runTimeLocale: Locale;
const changeRunTimeLocale = (locale: Locale) => {
  runTimeLocale = locale;
};

export const getRunTimeLocale = () => runTimeLocale;

export default class ConfigProvider extends PureComponent<ConfigProviderProps, {}> {
  static defaultProps = defaultConfig;

  componentDidMount() {
    this.update();
  }

  componentDidUpdate() {
    this.update();
  }

  update() {
    const { locale, theme, primaryColor } = this.props;
    changeRunTimeLocale(locale);
    setTheme(theme);
    setPrimaryColor(primaryColor);
  }

  render() {
    const { locale, theme, primaryColor, children } = this.props;
    return (
      <ConfigContext.Provider value={{ locale, theme, primaryColor }}>
        {React.Children.only(children)}
      </ConfigContext.Provider>
    );
  }
}
