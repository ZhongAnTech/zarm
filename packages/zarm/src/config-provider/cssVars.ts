import type { Replace } from '../utils/utilityTypes';
import type { ActionSheetCssVars } from '../action-sheet';
import type { ActivityIndicatorCssVars } from '../activity-indicator';
import type { AffixCssVars } from '../affix';
import type { BackTopCssVars } from '../back-top';
import type { BadgeCssVars } from '../badge';
import type { ButtonCssVars } from '../button';
import type { CalendarCssVars } from '../calendar';
import type { CarouselCssVars } from '../carousel';
import type { CheckboxCssVars, CheckboxGroupCssVars } from '../checkbox';
import type { CollapseCssVars } from '../collapse';
import type { CustomInputCssVars } from '../custom-input';
import type { GridCssVars } from '../grid';
import type { ImagePreviewCssVars } from '../image-preview';
import type { InputCssVars } from '../input';
import type { KeyboardCssVars } from '../keyboard';
import type { ListCssVars } from '../list';
import type { SliderCssVars } from '../slider';
import type { SwitchCssVars } from '../switch';
import type { ModalCssVars } from '../modal';
import type { IconCssVars } from '../icon';
import type { RateCssVars } from '../rate';
import type { PullCssVars } from '../pull';
import type { MaskCssVars } from '../mask';
import type { LoadingCssVars } from '../loading';
import type { MessageCssVars } from '../message';
import type { NavBarCssVars } from '../nav-bar';
import type { NoticeBarCssVars } from '../notice-bar';
import type { PanelCssVars } from '../panel';
import type { PickerViewCssVars } from '../picker-view';
import type { PickerCssVars } from '../picker';
import type { SelectCssVars } from '../select';
import type { ProgressCssVars } from '../progress';
import type { RadioGroupCssVars, RadioCssVars } from '../radio';
import type { SearchBarCssVars } from '../search-bar';
import type { SkeletonCssVars } from '../skeleton';
import type { StepperCssVars } from '../stepper';
import type { SwipeActionCssVars } from '../swipe-action';
import type { TabsCssVars } from '../tabs';
import type { TooltipCssVars } from '../tooltip';
import type { WheelCssVars } from '../wheel';
import type { TabBarCssVars } from '../tab-bar';
import type { ImageCssVars } from '../image';
import type { CascaderViewCssVars } from '../cascader-view';

type Prefix<T extends object, N extends string> = {
  [Key in keyof T as Key extends string ? `--za-${N}-${Replace<Key, '--', ''>}` : never]: T[Key];
};

export type CssVars = Prefix<ActionSheetCssVars, 'action-sheet'> &
  Prefix<ActivityIndicatorCssVars, 'activity-indicator'> &
  Prefix<AffixCssVars, 'affix'> &
  Prefix<GridCssVars, 'grid'> &
  Prefix<BackTopCssVars, 'back-top'> &
  Prefix<BadgeCssVars, 'badge'> &
  Prefix<ButtonCssVars, 'button'> &
  Prefix<CalendarCssVars, 'calendar'> &
  Prefix<CarouselCssVars, 'carousel'> &
  Prefix<CheckboxCssVars & CheckboxGroupCssVars, 'checkbox'> &
  Prefix<CollapseCssVars, 'collapse'> &
  Prefix<CustomInputCssVars, 'custom-input'> &
  Prefix<ImagePreviewCssVars, 'image-preview'> &
  Prefix<InputCssVars, 'input'> &
  Prefix<KeyboardCssVars, 'keyboard'> &
  Prefix<ListCssVars, 'list'> &
  Prefix<LoadingCssVars, 'loading'> &
  Prefix<SliderCssVars, 'slider'> &
  Prefix<SwitchCssVars, 'switch'> &
  Prefix<ModalCssVars, 'modal'> &
  Prefix<IconCssVars, 'icon'> &
  Prefix<RateCssVars, 'rate'> &
  Prefix<MaskCssVars, 'mask'> &
  Prefix<PullCssVars, 'pull'> &
  Prefix<MessageCssVars, 'message'> &
  Prefix<NavBarCssVars, 'nav-bar'> &
  Prefix<NoticeBarCssVars, 'notice-bar'> &
  Prefix<PanelCssVars, 'panel'> &
  Prefix<PickerViewCssVars, 'picker-view'> &
  Prefix<PickerCssVars, 'picker'> &
  Prefix<SelectCssVars, 'select'> &
  Prefix<ProgressCssVars, 'progress'> &
  Prefix<RadioCssVars & RadioGroupCssVars, 'radio'> &
  Prefix<SearchBarCssVars, 'search-bar'> &
  Prefix<SkeletonCssVars, 'skeleton'> &
  Prefix<StepperCssVars, 'stepper'> &
  Prefix<SwipeActionCssVars, 'swipe-action'> &
  Prefix<TabsCssVars, 'tabs'> &
  Prefix<TooltipCssVars, 'tooltip'> &
  Prefix<WheelCssVars, 'wheel'> &
  Prefix<TabBarCssVars, 'tab-bar'> &
  Prefix<ImageCssVars, 'image'> &
  Prefix<CascaderViewCssVars, 'cascader-view'>;
