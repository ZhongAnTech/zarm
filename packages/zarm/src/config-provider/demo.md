# ConfigProvider 全局配置

## 基本用法

```jsx
import { useState } from 'react';
import {
  ConfigProvider,
  List,
  Button,
  SearchBar,
  Modal,
  Keyboard,
  Radio,
  Message,
  TabBar,
  Icon,
  Popup,
} from 'zarm';
import enUS from 'zarm/lib/config-provider/locale/en_US';
import zhCN from 'zarm/lib/config-provider/locale/zh_CN';

const TabIcon = Icon.createFromIconfont('//at.alicdn.com/t/font_1340918_lpsswvb7yv.js');

const colors = ['#00bc70', '#1890ff', '#f5222d', '#fa541b', '#13c2c2', '#2f54ec', '#712fd1'];

const Demo = () => {
  const [locale, setLocale] = useState(localStorage.locale || 'zhCN');
  const [theme, setTheme] = useState(localStorage.theme || 'light');
  const [primaryColor, setPrimaryColor] = useState(localStorage.primaryColor || colors[0]);
  const [visiblePopup, setVisiblePopup] = useState(false);

  const show = (key) => {
    if (key === 'alert') {
      Modal.alert({
        title: '警告',
        content: '这里是警告信息',
        shape: 'radius',
      });
    } else {
      Modal.confirm({
        title: '确认信息',
        content: '你确定要这样做吗？',
        shape: 'radius',
      });
    }
  };

  return (
    <>
      <List>
        <List.Item
          title="切换语言"
          after={
            <Radio.Group buttonCompact type="button" value={locale} onChange={setLocale}>
              <Radio value="zhCN">中文</Radio>
              <Radio value="enUS">EN</Radio>
            </Radio.Group>
          }
        />
        <List.Item
          title="切换品牌色"
          after={
            <ul className="colors">
              {colors.map((color, index) => {
                return (
                  <li
                    key={+index}
                    style={{ backgroundColor: color }}
                    onClick={() => setPrimaryColor(color)}
                  />
                );
              })}
            </ul>
          }
        />
        <List.Item
          title="切换主题"
          after={
            <Radio.Group
              buttonCompact
              type="button"
              value={theme}
              onChange={(value) => setTheme(value)}
            >
              <Radio value="light">默认主题</Radio>
              <Radio value="dark">暗黑主题</Radio>
            </Radio.Group>
          }
        />
      </List>

      <Message theme="warning">以下为组件示例</Message>

      <ConfigProvider
        locale={locale === 'enUS' ? enUS : zhCN}
        primaryColor={primaryColor}
        theme={theme}
        safeIphoneX
      >
        <>
          <SearchBar />
          <Keyboard />

          <List>
            <List.Item
              after={
                <Button size="xs" onClick={() => show('alert')}>
                  开启
                </Button>
              }
            >
              警告框
            </List.Item>
            <List.Item
              after={
                <Button size="xs" onClick={() => show('confirm')}>
                  开启
                </Button>
              }
            >
              确认框
            </List.Item>
            <List.Item
              after={
                <Button size="xs" onClick={() => setVisiblePopup(true)}>
                  开启
                </Button>
              }
            >
              弹出层
            </List.Item>
          </List>
          <Popup
            visible={visiblePopup}
            direction="bottom"
            onMaskClick={() => setVisiblePopup(false)}
            afterOpen={() => console.log('打开')}
            afterClose={() => console.log('关闭')}
            className="provider-popup"
            destroy={false}
            mountContainer={() => document.body}
          >
            <div className="popup-box-bottom" onClick={() => setVisiblePopup(false)}>
              关闭
            </div>
            <Button block theme="primary">
              按钮
            </Button>
          </Popup>

          <TabBar visible={true} activeKey={1}>
            <TabBar.Item itemKey="home" title="主页" icon={<TabIcon type="home" />} />
            <TabBar.Item
              itemKey="found"
              title="保险"
              icon={<TabIcon type="insurance" />}
              badge={{ shape: 'circle', text: '3' }}
            />
            <TabBar.Item
              itemKey="me"
              title="我的"
              icon={<TabIcon type="user" />}
              badge={{ shape: 'dot' }}
            />
          </TabBar>
        </>
      </ConfigProvider>
    </>
  );
};

ReactDOM.render(<Demo />, mountNode);
```

## API

| 属性         | 类型              | 默认值    | 说明                                                                                                                                |
| :----------- | :---------------- | :-------- | :---------------------------------------------------------------------------------------------------------------------------------- |
| locale       | Object            | -         | 语言包配置，默认为中文，语言包可到 [zarm/lib/config-provider/locale](https://unpkg.com/zarm/lib/config-provider/locale/) 目录下寻找 |
| theme        | 'light' \| 'dark' | 'light'   | 主题模式，光亮主题 和 暗黑主题的切换                                                                                                |
| primaryColor | string            | '#00bc70' | 品牌标准色                                                                                                                          |
| safeIphoneX  | boolean           | false     | 是否适配 iphoneX 刘海屏                                                                                                             |
