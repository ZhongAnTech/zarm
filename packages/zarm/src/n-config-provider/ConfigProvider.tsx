import * as React from 'react';
import type { Context } from 'react';
import type { ConfigProviderProps } from './interface';
import defaultLocaleData from './locale/zh_CN';
import setCssVars, { defaultCssVars } from './setCssVars';
import setPrimaryColor from './setPrimaryColor';
import setTheme from './setTheme';

const defaultConfig: ConfigProviderProps = {
  prefixCls: 'za',
  locale: defaultLocaleData,
  theme: 'light',
  safeIphoneX: false,
  cssVars: defaultCssVars,
};

export const ConfigContext: Context<ConfigProviderProps> = React.createContext(defaultConfig);

let runtimeConfigContext: ConfigProviderProps;

const changeRuntimeConfigContext = (props: ConfigProviderProps) => {
  runtimeConfigContext = props;
};

const ConfigProvider: React.FC<ConfigProviderProps> = (props) => {
  const { children, cssVars, primaryColor, theme, ...rest } = props;

  changeRuntimeConfigContext(props);

  React.useEffect(() => {
    primaryColor && setPrimaryColor(primaryColor);
  }, [primaryColor]);

  React.useEffect(() => {
    theme && setTheme(theme);
  }, [theme]);

  const newChildren = setCssVars(children, cssVars);

  return (
    <ConfigContext.Provider value={rest}>{React.Children.only(newChildren)}</ConfigContext.Provider>
  );
};

export const RuntimeConfigProvider: React.FC = ({ children }) => {
  const props = React.useRef(runtimeConfigContext);

  React.useEffect(() => {
    props.current = runtimeConfigContext;
  }, [runtimeConfigContext]);

  return <ConfigProvider {...props?.current}>{children}</ConfigProvider>;
};

ConfigProvider.displayName = 'ConfigProvider';
ConfigProvider.defaultProps = defaultConfig;

export default ConfigProvider;
