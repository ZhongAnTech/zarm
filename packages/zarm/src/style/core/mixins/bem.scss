@import '../functions/string';

$--namespace: 'za';
$--block-separator: '-';
$--element-separator: '__';
$--modifier-separator: '--';
$--keyframes-separator: '_';

// 是否包含Modifier
@function contains-modifier($selector) {
  @return str-contains(str-selector($selector), $--modifier-separator);
}

// 是否包含Element
@function containsElement($selector) {
  @return str-contains(str-selector($selector), $--element-separator);
}

// 获取块名
@function get-block($selector) {
  $selector: str-selector($selector);
  $parent: str-slice($selector, str-index-reverse($selector, '.'));
  $start: -2;
  @if containsElement($parent) {
    $start: str-index($parent, $--element-separator) - 1;
  }
  @else if contains-modifier($parent) {
    $start: str-index($parent, $--modifier-separator) - 1;
  }
  $block: str-slice($parent, 0, $start);
  // @debug $block;
  @return $block;
}

// 获取节点列表
@function get-elements($elements) {
  $result: '';
  $block: get-block(&);
  $selector: if(contains-modifier(&), get-block(&), &);
  @each $unit in $elements {
    $element: selector-append($selector, $--element-separator + $unit);
    @if containsElement($selector) {
      $element: selector-nest($selector, selector-append($block, $--element-separator + $unit));
    }
    $result: $result + $element + ',';
  }
  @return $result;
}

/**
 * block define
 * @param $name     block name
 * @examples
 *
 *    @include b(button) { ... }
 *    =>
 *    .za-button { ... }
 *
 *
 *    @include b(checkbox-group) { ... }
 *    =>
 *    .za-checkbox-group { ... }
 */
@mixin b($block) {
  $block: selector-append($--namespace, $--block-separator + $block);
  // @debug $block;
  .#{$block} {
    @content;
  }
}

/**
 * element define
 * @param $list     list of element names
 * @examples
 *
 *    @include b(button) {
 *      @include e(icon) { ... }
 *    }
 *    =>
 *    .za-button__icon { ... }
 *
 *
 *    @include b(dialog) {
 *      @include e(header, footer) { ... }
 *    }
 *    =>
 *    .za-dialog__header, .za-dialog__footer { ... }
 */
@mixin e($elements...) {
  $elements: get-elements($elements);
  $selector: if(contains-modifier(&), selector-nest(&, $elements), $elements);
  @at-root {
    #{$selector} {
      @content;
    }
  }
}

/**
 * modifier define
 * @param $modifier     modifier name
 * @examples
 *
 *    @include b(button) {
 *      @include m(danger) { ... }
 *    }
 *    =>
 *    .za-button--danger { ... }
 *
 *
 *    @include b(button) {
 *      @include m(danger) {
 *        @include e(content) { ... }
 *      }
 *    }
 *    =>
 *    .za-button--danger .za-button__content { ... }
 *
 *
 *    @include b(button) {
 *      @include e(content) {
 *        @include m(danger) { ... }
 *      }
 *    }
 *    =>
 *    ..za-button__content--danger { ... }
 */
@mixin m($modifier) {
  $parent: if(contains-modifier(&), & + get-block(&), &);
  $selector: selector-append($parent, $--modifier-separator + $modifier);
  @at-root {
    #{$selector} {
      @content;
    }
  }
}
