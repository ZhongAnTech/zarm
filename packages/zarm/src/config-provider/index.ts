import ConfigProvider from './ConfigProvider';

export type { ConfigProviderProps } from './interface';

export { ConfigContext } from './ConfigProvider';
export default ConfigProvider;
