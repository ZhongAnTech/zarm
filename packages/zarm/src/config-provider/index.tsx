import ConfigProvider from './ConfigProvider';

export default ConfigProvider;
