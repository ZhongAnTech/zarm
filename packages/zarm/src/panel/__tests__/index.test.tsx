import { render } from '@testing-library/react';
import { createBEM } from '@zarm-design/bem';
import React from 'react';
import { defaultConfig } from '../../config-provider/ConfigProvider';
import Panel from '../index';

const bem = createBEM('panel', { prefixCls: defaultConfig.prefixCls });

describe('Panel', () => {
  test('snapshots with title and more props', () => {
    const { asFragment } = render(
      <Panel title="title" more="more">
        body
      </Panel>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  test('should not render header if do not have title and more props', () => {
    const { container } = render(<Panel>body</Panel>);
    expect(container.children).toHaveLength(1);
    expect(container.querySelector(`.${bem('header')}`)).not.toBeInTheDocument();
    expect(container.querySelector(`.${bem('body')}`)).toBeInTheDocument();
  });

  test('should accept custom css vars', () => {
    const { container } = render(
      <Panel title="title" style={{ '--header-padding': 0 }}>
        body
      </Panel>,
    );
    expect(container.querySelector(`.${bem('header')}`)).toHaveStyle({
      padding: 'var(--header-padding)',
    });
  });
});
