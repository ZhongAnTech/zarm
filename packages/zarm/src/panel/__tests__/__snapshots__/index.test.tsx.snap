// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel snapshots with title and more props 1`] = `
<div
  class="za-panel za-panel--bordered"
>
  <div
    class="za-panel__header"
  >
    <div
      class="za-panel__header__title"
    >
      title
    </div>
    <div
      class="za-panel__header__more"
    >
      more
    </div>
  </div>
  <div
    class="za-panel__body"
  >
    body
  </div>
</div>
`;
